### PR TITLE
adding init_cookies function

### PR DESCRIPTION
**Pull request by jen-machin**
Added init_cookies function and tried to account for scenarios where:

- JS file already exists
- JS file doesn't exist
- Download fails for some other reason (e.g. URL broken/user not connected to internet) 

I've tested each outcome and it does seem to work but I can easily be wrong!

### CODE REVIEW

**Comment by jen-machin** [comment]
Have added in a function that creates the www/ folder if it doesn't already exist

File: R/cookies.R [hunk 1]
```diff
@@ -246,3 +247,43 @@ cookie_banner_server <- function(
...
+#' It checks whether there is already a www/ folder and if not, it creates one
+#' It then checks whether the cookie-consent.js file exists in the www/ folder
+#' If the file exists, it will print a message in the console to let you know
+#' If the file doesn't exist, it will pull a copy from the GitHub repo
+#' If it cannot connect to the repo then it will print "Download failed".
+#' No input parameters are required
+#'
+#' Call init_cookies() in the console to run the function
+#'
+
```
Comment (rmbielby) [inline R/cookies.R:264]: I'd prefer not to have the whitespace lines here, looks a bit confusing like the roxygen comments aren't associated with the function.

File: R/cookies.R [hunk 2]
```diff
@@ -246,3 +247,43 @@ cookie_banner_server <- function(
...
+
```
Comment (rmbielby) [inline R/cookies.R:265]: init_cookies currently isn't exported as a user function, so won't work when people load the package. Need to add the @export line to the roxygen comments (can check the other functions for how it needs to be done).

File: R/cookies.R [hunk 3]
```diff
@@ -246,3 +247,43 @@ cookie_banner_server <- function(
...
+#' @description
+#' init_cookies() creates a local copy of the JavaScript file required for cookies to work.
+#' It checks whether there is already a www/ folder and if not, it creates one
+#' It then checks whether the cookie-consent.js file exists in the www/ folder
+#' If the file exists, it will print a message in the console to let you know
+#' If the file doesn't exist, it will pull a copy from the GitHub repo
+#' If it cannot connect to the repo then it will print "Download failed".
+#' No input parameters are required
+#'
+#' Call init_cookies() in the console to run the function
```
Comment (cjrace) [inline R/cookies.R:262]: For the example I'd do something like this probably so it shows in the standard way on the documentation
 
#' @examples
#' if (interactive()) {
#'   init_cookies()
#' }

**Review by cjrace — CHANGES_REQUESTED** [review]
I think you're missing some of the standard roxygen sections, have you ran devtools::check() locally to test the package on this branch? 

That should hopefully tell you the ones missing, off the top of my head, I know it usually expects @examples to present.

I think the checks won't run on the PR itself until you resolve the conflicts that exist too, so have a go at removing them and we'll be able to see if anything else comes up in the automated checks

**Comment by cjrace** [comment]
[Finish init_cookies documentation](https://trello.com/c/y4UXBmwk/1359-finish-initcookies-documentation)

File: R/cookies.R [hunk 4]
```diff
@@ -209,3 +209,43 @@
...
+  output_dir <- file.path(sub_dir)
+
+  if (!dir.exists(output_dir)){
+    dir.create(output_dir)
+  } else {
+    print("www folder already exists!")
+  }
+
+    tryCatch(
+      download.file(url = "https://raw.githubusercontent.com/dfe-analytical-services/dfeshiny/main/inst/cookie-consent.js", destfile = "www/cookie-consent.js"),
```
[the remark this answers is not in the document]
Comment (cjrace) [reply R/cookies.R:245]: As this is just a long url, I'd stick `# nolint: [line_length_linter]` at the end of the line and that will tell lintr to ignore the line length being long here.

It's either that or cut up the URL into chunks and paste together but that feels excessive

File: R/cookies.R [hunk 5]
```diff
@@ -209,3 +209,43 @@
...
+#' if (interactive()) {
+#' init_cookies()
+#' }
+init_cookies <- function() {
+
+  sub_dir <- "www"
+
+  output_dir <- file.path(sub_dir)
+
+  if (!dir.exists(output_dir)){
```
[the remark this answers is not in the document]
Comment (cjrace) [reply R/cookies.R:238]: Might be worth running `styler::style_pkg()` locally to see if it'll automatically clean this up?

[the remark this answers is not in the document]
Comment (cjrace) [reply R/cookies.R:238]: Might be worth running styler::style_pkg() locally to see if it'll automatically clean this up?

File: R/cookies.R [hunk 6]
```diff
@@ -209,3 +209,43 @@
     }))
   })
 }
+
+#' init_cookies
+#'
+#' @description
+#' init_cookies() creates a local copy of the JavaScript file required for cookies to work.
```
[the remark this answers is not in the document]
Comment (cjrace) [reply R/cookies.R:216]: I'd break this over two lines to hush the alert, should still render the same even if split over multiple lines

File: R/cookies.R [hunk 7]
```diff
@@ -209,3 +209,43 @@
...
+
+  output_dir <- file.path(sub_dir)
+
+  if (!dir.exists(output_dir)){
+    dir.create(output_dir)
+  } else {
+    print("www folder already exists!")
+  }
+
+    tryCatch(
```
[the remark this answers is not in the document]
Comment (cjrace) [reply R/cookies.R:244]: Might be worth running styler::style_pkg() locally to see if it'll automatically clean this up?

File: R/cookies.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jen-machin) [reply R/cookies.R:245]: 🙃

File: R/cookies.R [hunk 8]
```diff
@@ -203,190 +203,236 @@ cookie_banner_server <- function(
...
 
-#' cookies_panel_ui
+#' init_cookies
 #'
 #' @description
-#' Create the standard DfE R-Shiny cookies dashboard panel in the ui. The server
-#' side functionality is provided by cookies_panel_server()
+#' init_cookies() creates a local copy of the JavaScript file
+#' required for cookies to work.
+#' It checks whether there is already a www/ folder and if not, it creates one
```
Comment (rmbielby) [inline R/cookies.R:211]: Can you add full-stops at the end of these lines. When it renders in a help page, it takes out the carriage returns and just flows on, which makes it awkward to read if there's no full-stops at the end of sentences.

File: R/cookies.R [hunk 9]
```diff
@@ -203,190 +203,236 @@ cookie_banner_server <- function(
...
+  } else {
+    print("www folder already exists!")
+  }
+
+  tryCatch(
+    download.file(url = "https://raw.githubusercontent.com/dfe-analytical-services/dfeshiny/main/inst/cookie-consent.js", destfile = "www/cookie-consent.js"), # nolint: [line_length_linter]
+    error = function(e) {
+      return("Download failed")
+    },
+    print("Cookie script updated")
```
Comment (rmbielby) [inline R/cookies.R:242]: `message()` probably better to use than `print()` here.

File: R/cookies.R [hunk 10]
```diff
@@ -203,190 +203,236 @@ cookie_banner_server <- function(
...
-          shiny::tags$p("Cookies are small files saved on your phone, tablet or
+init_cookies <- function() {
+  sub_dir <- "www"
+
+  output_dir <- file.path(sub_dir)
+
+  if (!dir.exists(output_dir)) {
+    dir.create(output_dir)
+  } else {
+    print("www folder already exists!")
```
Comment (rmbielby) [inline R/cookies.R:234]: Probably better to use `message()` in place of `print()`.

**Review by rmbielby — CHANGES_REQUESTED** [review]
Tested it within the package and loaded to another project and looks like it works well.

Handful of minor changes requested.

**Review by rmbielby — APPROVED** [review]

**Review by cjrace — APPROVED** [review]
As mentioned on Trello, happy to approve given Rich is happy and I'll do a review of the examples and documentation in my follow up PR anyway too

**Pull request merged**